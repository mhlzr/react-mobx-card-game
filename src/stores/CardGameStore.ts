import { observable, flow, action, computed } from 'mobx';
import { Player } from '../models/Player';
import { fetchPlayers, savePlayerSelection } from '../api/players';

export enum PLAYER_SORTATION {
    ASCENDING,
    DESCENDING
}

export class CardGameStore {

    @observable _players: Array<Player> = [];

    @observable playerSortation: PLAYER_SORTATION | null = null;
    @observable player: Player | null = null;

    @observable isFetching: Boolean = false;
    @observable isSaving: Boolean = false;


    fetchPlayers = flow(function* () {
        this.isFetching = true;

        try {
            this._players = yield fetchPlayers();
        }
        catch (error) {
            console.error('Could not fetch players');
        }
        finally {
            this.isFetching = false;
        }
    });

    savePlayerSelection = flow(function* () {
        this.isSaving = true;

        try {
            const response = yield savePlayerSelection();
            console.log('Done');
        }
        catch (error) {
            console.error('Could not save playerSelection');
        }
        finally {
            this.isSaving = false;
        }
    });

    @computed get players() {
        switch (this.playerSortation) {
            case PLAYER_SORTATION.ASCENDING: return this._players.slice().sort(); break;
            case PLAYER_SORTATION.DESCENDING: return this._players.slice().sort().reverse(); break;
            default: return this._players;
        }
    }
}

export default new CardGameStore();
