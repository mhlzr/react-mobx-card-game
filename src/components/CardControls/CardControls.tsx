import React, { ReactElement, FunctionComponent } from 'react';
import styled from 'styled-components';
import { observer, inject } from 'mobx-react';

import { Headline } from '../Headline/Headline';
import { Button } from '../Button/Button';
import { CardGameStore, PLAYER_SORTATION } from '../../stores/CardGameStore';
import { LoadingSpinner } from '../LoadingSpinner/LoadingSpinner';

const Section = styled.section`
    border: 1px solid var(--secondary-color);
    grid-area: controls;
    padding: 1em;
`;

const StyledButton = styled(Button)`
    text-transform: uppercase;

    &:last-child {
        grid-column: 1 / 3; 
    }
`;

const Controls = styled.div`
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-gap: 1em;
    margin-top: 1em;
`;

interface CardControlsProps {
    store?: CardGameStore
}

export const CardControlsComponent: FunctionComponent<CardControlsProps> = ({ store = new CardGameStore() }: CardControlsProps): ReactElement => {
    const { player, isSaving } = store;

    const onSortAscendingClick = () => {
        store.playerSortation = PLAYER_SORTATION.ASCENDING;
    }

    const onSortDescdendingClick = () => {
        store.playerSortation = PLAYER_SORTATION.DESCENDING;
    }

    const onSubmitClick = () => {
        store.savePlayerSelection();
    };

    return (
        <Section>
            <Headline>Controls</Headline>
            <Controls>
                <StyledButton onClick={onSortAscendingClick} selected={store.playerSortation === PLAYER_SORTATION.ASCENDING}>Sort Asc</StyledButton>
                <StyledButton onClick={onSortDescdendingClick} selected={store.playerSortation === PLAYER_SORTATION.DESCENDING}>Sort Desc</StyledButton>
                <StyledButton disabled={!player || isSaving} onClick={onSubmitClick}>
                    {isSaving && <LoadingSpinner />}
                    {!isSaving && 'Submit'}
                </StyledButton>
            </Controls>
        </Section>
    )
};

export const CardControls = inject('store')(observer(CardControlsComponent));
