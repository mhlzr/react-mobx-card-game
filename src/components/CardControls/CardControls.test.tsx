import React from 'react';
import { shallow } from 'enzyme'

import { CardControls } from './CardControls';
import { CardGameStore } from '../../stores/CardGameStore';

describe('CardControls', () => {

    let store: CardGameStore;

    beforeEach(() => {
        store = new CardGameStore();
    });

    it('does not render any content if there is no player selected', () => {
        store.player = null;
        const wrapper = shallow(<CardControls store={store} />);
        expect(wrapper.children()).toHaveLength(0)
    });

    it('has a snapshot', () => {
        const wrapper = shallow(<CardControls store={store} />);
        expect(wrapper).toMatchSnapshot();
    });
})
