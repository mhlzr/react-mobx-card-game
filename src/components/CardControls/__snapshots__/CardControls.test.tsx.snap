// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardControls has a snapshot 1`] = `ShallowWrapper {}`;
